// osmojs/module/codegen/osmosis/concentrated-liquidity/pool-model/query.lcd.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { setPaginationParams } from "../../../helpers";
export class LCDQueryClient {
  constructor({
    requestClient
  }) {
    _defineProperty(this, "req", void 0);

    this.req = requestClient;
    this.pools = this.pools.bind(this);
    this.pool = this.pool.bind(this);
    this.params = this.params.bind(this);
  }
  /* Pools returns all concentrated liquidity pools */


  async pools(params = {
    pagination: undefined
  }) {
    const options = {
      params: {}
    };

    if (typeof (params === null || params === void 0 ? void 0 : params.pagination) !== "undefined") {
      setPaginationParams(options, params.pagination);
    }

    const endpoint = `osmosis/concentratedliquidity/v1beta1/pools`;
    return await this.req.get(endpoint, options);
  }
  /* Pool returns the Pool specified by the pool id */


  async pool(params) {
    const endpoint = `osmosis/concentratedliquidity/v1beta1/pools/${params.poolId}`;
    return await this.req.get(endpoint);
  }
  /* Params returns concentrated liquidity module params. */


  async params(_params = {}) {
    const endpoint = `osmosis/concentratedliquidity/v1beta1/params`;
    return await this.req.get(endpoint);
  }

}
