// osmojs/module/codegen/osmosis/concentrated-liquidity/pool-model/tx.js
import * as _m0 from "protobufjs/minimal";
import { Long } from "../../../helpers";
/** ===================== MsgCreateConcentratedPool */

function createBaseMsgCreateConcentratedPool() {
  return {
    sender: "",
    denom0: "",
    denom1: "",
    tickSpacing: Long.UZERO,
    precisionFactorAtPriceOne: "",
    swapFee: ""
  };
}

export const MsgCreateConcentratedPool = {
  encode(message, writer = _m0.Writer.create()) {
    if (message.sender !== "") {
      writer.uint32(10).string(message.sender);
    }

    if (message.denom0 !== "") {
      writer.uint32(18).string(message.denom0);
    }

    if (message.denom1 !== "") {
      writer.uint32(26).string(message.denom1);
    }

    if (!message.tickSpacing.isZero()) {
      writer.uint32(32).uint64(message.tickSpacing);
    }

    if (message.precisionFactorAtPriceOne !== "") {
      writer.uint32(42).string(message.precisionFactorAtPriceOne);
    }

    if (message.swapFee !== "") {
      writer.uint32(74).string(message.swapFee);
    }

    return writer;
  },

  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseMsgCreateConcentratedPool();

    while (reader.pos < end) {
      const tag = reader.uint32();

      switch (tag >>> 3) {
        case 1:
          message.sender = reader.string();
          break;

        case 2:
          message.denom0 = reader.string();
          break;

        case 3:
          message.denom1 = reader.string();
          break;

        case 4:
          message.tickSpacing = reader.uint64();
          break;

        case 5:
          message.precisionFactorAtPriceOne = reader.string();
          break;

        case 9:
          message.swapFee = reader.string();
          break;

        default:
          reader.skipType(tag & 7);
          break;
      }
    }

    return message;
  },

  fromPartial(object) {
    var _object$sender, _object$denom, _object$denom2, _object$precisionFact, _object$swapFee;

    const message = createBaseMsgCreateConcentratedPool();
    message.sender = (_object$sender = object.sender) !== null && _object$sender !== void 0 ? _object$sender : "";
    message.denom0 = (_object$denom = object.denom0) !== null && _object$denom !== void 0 ? _object$denom : "";
    message.denom1 = (_object$denom2 = object.denom1) !== null && _object$denom2 !== void 0 ? _object$denom2 : "";
    message.tickSpacing = object.tickSpacing !== undefined && object.tickSpacing !== null ? Long.fromValue(object.tickSpacing) : Long.UZERO;
    message.precisionFactorAtPriceOne = (_object$precisionFact = object.precisionFactorAtPriceOne) !== null && _object$precisionFact !== void 0 ? _object$precisionFact : "";
    message.swapFee = (_object$swapFee = object.swapFee) !== null && _object$swapFee !== void 0 ? _object$swapFee : "";
    return message;
  }

};

function createBaseMsgCreateConcentratedPoolResponse() {
  return {
    poolId: Long.UZERO
  };
}

export const MsgCreateConcentratedPoolResponse = {
  encode(message, writer = _m0.Writer.create()) {
    if (!message.poolId.isZero()) {
      writer.uint32(8).uint64(message.poolId);
    }

    return writer;
  },

  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseMsgCreateConcentratedPoolResponse();

    while (reader.pos < end) {
      const tag = reader.uint32();

      switch (tag >>> 3) {
        case 1:
          message.poolId = reader.uint64();
          break;

        default:
          reader.skipType(tag & 7);
          break;
      }
    }

    return message;
  },

  fromPartial(object) {
    const message = createBaseMsgCreateConcentratedPoolResponse();
    message.poolId = object.poolId !== undefined && object.poolId !== null ? Long.fromValue(object.poolId) : Long.UZERO;
    return message;
  }

};
