// osmojs/main/codegen/osmosis/concentrated-liquidity/pool.js
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Pool = void 0;

var _m0 = _interopRequireWildcard(require("protobufjs/minimal"));

var _helpers = require("../../helpers");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function createBasePool() {
  return {
    address: "",
    id: _helpers.Long.UZERO,
    liquidity: "",
    token0: "",
    token1: "",
    currentSqrtPrice: "",
    currentTick: "",
    tickSpacing: _helpers.Long.UZERO,
    precisionFactorAtPriceOne: "",
    swapFee: ""
  };
}

var Pool = {
  encode: function encode(message) {
    var writer = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _m0.Writer.create();

    if (message.address !== "") {
      writer.uint32(10).string(message.address);
    }

    if (!message.id.isZero()) {
      writer.uint32(16).uint64(message.id);
    }

    if (message.liquidity !== "") {
      writer.uint32(26).string(message.liquidity);
    }

    if (message.token0 !== "") {
      writer.uint32(34).string(message.token0);
    }

    if (message.token1 !== "") {
      writer.uint32(42).string(message.token1);
    }

    if (message.currentSqrtPrice !== "") {
      writer.uint32(50).string(message.currentSqrtPrice);
    }

    if (message.currentTick !== "") {
      writer.uint32(58).string(message.currentTick);
    }

    if (!message.tickSpacing.isZero()) {
      writer.uint32(64).uint64(message.tickSpacing);
    }

    if (message.precisionFactorAtPriceOne !== "") {
      writer.uint32(74).string(message.precisionFactorAtPriceOne);
    }

    if (message.swapFee !== "") {
      writer.uint32(82).string(message.swapFee);
    }

    return writer;
  },
  decode: function decode(input, length) {
    var reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    var end = length === undefined ? reader.len : reader.pos + length;
    var message = createBasePool();

    while (reader.pos < end) {
      var tag = reader.uint32();

      switch (tag >>> 3) {
        case 1:
          message.address = reader.string();
          break;

        case 2:
          message.id = reader.uint64();
          break;

        case 3:
          message.liquidity = reader.string();
          break;

        case 4:
          message.token0 = reader.string();
          break;

        case 5:
          message.token1 = reader.string();
          break;

        case 6:
          message.currentSqrtPrice = reader.string();
          break;

        case 7:
          message.currentTick = reader.string();
          break;

        case 8:
          message.tickSpacing = reader.uint64();
          break;

        case 9:
          message.precisionFactorAtPriceOne = reader.string();
          break;

        case 10:
          message.swapFee = reader.string();
          break;

        default:
          reader.skipType(tag & 7);
          break;
      }
    }

    return message;
  },
  fromPartial: function fromPartial(object) {
    var _object$address, _object$liquidity, _object$token, _object$token2, _object$currentSqrtPr, _object$currentTick, _object$precisionFact, _object$swapFee;

    var message = createBasePool();
    message.address = (_object$address = object.address) !== null && _object$address !== void 0 ? _object$address : "";
    message.id = object.id !== undefined && object.id !== null ? _helpers.Long.fromValue(object.id) : _helpers.Long.UZERO;
    message.liquidity = (_object$liquidity = object.liquidity) !== null && _object$liquidity !== void 0 ? _object$liquidity : "";
    message.token0 = (_object$token = object.token0) !== null && _object$token !== void 0 ? _object$token : "";
    message.token1 = (_object$token2 = object.token1) !== null && _object$token2 !== void 0 ? _object$token2 : "";
    message.currentSqrtPrice = (_object$currentSqrtPr = object.currentSqrtPrice) !== null && _object$currentSqrtPr !== void 0 ? _object$currentSqrtPr : "";
    message.currentTick = (_object$currentTick = object.currentTick) !== null && _object$currentTick !== void 0 ? _object$currentTick : "";
    message.tickSpacing = object.tickSpacing !== undefined && object.tickSpacing !== null ? _helpers.Long.fromValue(object.tickSpacing) : _helpers.Long.UZERO;
    message.precisionFactorAtPriceOne = (_object$precisionFact = object.precisionFactorAtPriceOne) !== null && _object$precisionFact !== void 0 ? _object$precisionFact : "";
    message.swapFee = (_object$swapFee = object.swapFee) !== null && _object$swapFee !== void 0 ? _object$swapFee : "";
    return message;
  }
};
exports.Pool = Pool;
