// osmojs/main/codegen/osmosis/concentrated-liquidity/tx.registry.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registry = exports.load = exports.MessageComposer = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _tx = require("./tx");

//@ts-nocheck
var registry = [["/osmosis.concentratedliquidity.v1beta1.MsgCreatePosition", _tx.MsgCreatePosition], ["/osmosis.concentratedliquidity.v1beta1.MsgWithdrawPosition", _tx.MsgWithdrawPosition], ["/osmosis.concentratedliquidity.v1beta1.MsgCollectFees", _tx.MsgCollectFees]];
exports.registry = registry;

var load = function load(protoRegistry) {
  registry.forEach(function (_ref) {
    var _ref2 = (0, _slicedToArray2["default"])(_ref, 2),
        typeUrl = _ref2[0],
        mod = _ref2[1];

    protoRegistry.register(typeUrl, mod);
  });
};

exports.load = load;
var MessageComposer = {
  encoded: {
    createPosition: function createPosition(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgCreatePosition",
        value: _tx.MsgCreatePosition.encode(value).finish()
      };
    },
    withdrawPosition: function withdrawPosition(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgWithdrawPosition",
        value: _tx.MsgWithdrawPosition.encode(value).finish()
      };
    },
    collectFees: function collectFees(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgCollectFees",
        value: _tx.MsgCollectFees.encode(value).finish()
      };
    }
  },
  withTypeUrl: {
    createPosition: function createPosition(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgCreatePosition",
        value: value
      };
    },
    withdrawPosition: function withdrawPosition(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgWithdrawPosition",
        value: value
      };
    },
    collectFees: function collectFees(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgCollectFees",
        value: value
      };
    }
  },
  fromPartial: {
    createPosition: function createPosition(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgCreatePosition",
        value: _tx.MsgCreatePosition.fromPartial(value)
      };
    },
    withdrawPosition: function withdrawPosition(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgWithdrawPosition",
        value: _tx.MsgWithdrawPosition.fromPartial(value)
      };
    },
    collectFees: function collectFees(value) {
      return {
        typeUrl: "/osmosis.concentratedliquidity.v1beta1.MsgCollectFees",
        value: _tx.MsgCollectFees.fromPartial(value)
      };
    }
  }
};
exports.MessageComposer = MessageComposer;
