// osmojs/module/codegen/osmosis/protorev/v1beta1/tx.amino.js
//@ts-nocheck
import { Long } from "../../../helpers";
export const AminoConverter = {
  "/osmosis.protorev.v1beta1.MsgSetHotRoutes": {
    aminoType: "osmosis/protorev/set-hot-routes",
    toAmino: ({
      admin,
      hotRoutes
    }) => {
      return {
        admin,
        hot_routes: hotRoutes.map(el0 => ({
          arb_routes: el0.arbRoutes.map(el1 => ({
            trades: el1.trades.map(el2 => ({
              pool: el2.pool.toString(),
              token_in: el2.tokenIn,
              token_out: el2.tokenOut
            }))
          })),
          token_in: el0.tokenIn,
          token_out: el0.tokenOut
        }))
      };
    },
    fromAmino: ({
      admin,
      hot_routes
    }) => {
      return {
        admin,
        hotRoutes: hot_routes.map(el0 => ({
          arbRoutes: el0.arb_routes.map(el1 => ({
            trades: el1.trades.map(el2 => ({
              pool: Long.fromString(el2.pool),
              tokenIn: el2.token_in,
              tokenOut: el2.token_out
            }))
          })),
          tokenIn: el0.token_in,
          tokenOut: el0.token_out
        }))
      };
    }
  },
  "/osmosis.protorev.v1beta1.MsgSetDeveloperAccount": {
    aminoType: "osmosis/protorev/set-developer-account",
    toAmino: ({
      admin,
      developerAccount
    }) => {
      return {
        admin,
        developer_account: developerAccount
      };
    },
    fromAmino: ({
      admin,
      developer_account
    }) => {
      return {
        admin,
        developerAccount: developer_account
      };
    }
  }
};
