// osmojs/main/codegen/osmosis/valset-pref/v1beta1/tx.rpc.msg.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MsgClientImpl = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _m0 = _interopRequireWildcard(require("protobufjs/minimal"));

var _tx = require("./tx");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var MsgClientImpl = /*#__PURE__*/function () {
  function MsgClientImpl(rpc) {
    (0, _classCallCheck2["default"])(this, MsgClientImpl);
    (0, _defineProperty2["default"])(this, "rpc", void 0);
    this.rpc = rpc;
    this.setValidatorSetPreference = this.setValidatorSetPreference.bind(this);
    this.delegateToValidatorSet = this.delegateToValidatorSet.bind(this);
    this.undelegateFromValidatorSet = this.undelegateFromValidatorSet.bind(this);
    this.redelegateValidatorSet = this.redelegateValidatorSet.bind(this);
    this.withdrawDelegationRewards = this.withdrawDelegationRewards.bind(this);
    this.delegateBondedTokens = this.delegateBondedTokens.bind(this);
  }

  (0, _createClass2["default"])(MsgClientImpl, [{
    key: "setValidatorSetPreference",
    value: function setValidatorSetPreference(request) {
      var data = _tx.MsgSetValidatorSetPreference.encode(request).finish();

      var promise = this.rpc.request("osmosis.valsetpref.v1beta1.Msg", "SetValidatorSetPreference", data);
      return promise.then(function (data) {
        return _tx.MsgSetValidatorSetPreferenceResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "delegateToValidatorSet",
    value: function delegateToValidatorSet(request) {
      var data = _tx.MsgDelegateToValidatorSet.encode(request).finish();

      var promise = this.rpc.request("osmosis.valsetpref.v1beta1.Msg", "DelegateToValidatorSet", data);
      return promise.then(function (data) {
        return _tx.MsgDelegateToValidatorSetResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "undelegateFromValidatorSet",
    value: function undelegateFromValidatorSet(request) {
      var data = _tx.MsgUndelegateFromValidatorSet.encode(request).finish();

      var promise = this.rpc.request("osmosis.valsetpref.v1beta1.Msg", "UndelegateFromValidatorSet", data);
      return promise.then(function (data) {
        return _tx.MsgUndelegateFromValidatorSetResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "redelegateValidatorSet",
    value: function redelegateValidatorSet(request) {
      var data = _tx.MsgRedelegateValidatorSet.encode(request).finish();

      var promise = this.rpc.request("osmosis.valsetpref.v1beta1.Msg", "RedelegateValidatorSet", data);
      return promise.then(function (data) {
        return _tx.MsgRedelegateValidatorSetResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "withdrawDelegationRewards",
    value: function withdrawDelegationRewards(request) {
      var data = _tx.MsgWithdrawDelegationRewards.encode(request).finish();

      var promise = this.rpc.request("osmosis.valsetpref.v1beta1.Msg", "WithdrawDelegationRewards", data);
      return promise.then(function (data) {
        return _tx.MsgWithdrawDelegationRewardsResponse.decode(new _m0.Reader(data));
      });
    }
  }, {
    key: "delegateBondedTokens",
    value: function delegateBondedTokens(request) {
      var data = _tx.MsgDelegateBondedTokens.encode(request).finish();

      var promise = this.rpc.request("osmosis.valsetpref.v1beta1.Msg", "DelegateBondedTokens", data);
      return promise.then(function (data) {
        return _tx.MsgDelegateBondedTokensResponse.decode(new _m0.Reader(data));
      });
    }
  }]);
  return MsgClientImpl;
}();

exports.MsgClientImpl = MsgClientImpl;
