// @osmonauts/ast/main/encoding/proto/to-amino/utils.js
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toAminoMessages = exports.toAminoJSON = exports.arrayTypes = void 0;
var t = _interopRequireWildcard(require("@babel/types"));
var _utils = require("../../../utils");
var _types2 = require("../../types");
var _implements = require("../implements");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var notUndefinedSetValue = function notUndefinedSetValue(sdkName, msgName, expr) {
  return t.expressionStatement(t.logicalExpression('&&', t.binaryExpression('!==', t.memberExpression(t.identifier('message'), t.identifier(msgName)), t.identifier('undefined')), t.assignmentExpression('=', t.memberExpression(t.identifier('obj'), t.identifier(sdkName)), expr)));
};
var toAminoJSON = {
  scalar: function scalar(args) {
    var _getFieldNames = (0, _types2.getFieldNames)(args.field),
      propName = _getFieldNames.propName,
      origName = _getFieldNames.origName;
    return t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('obj'), t.identifier(origName)), t.memberExpression(t.identifier('message'), t.identifier(propName))));
  },
  string: function string(args) {
    return toAminoJSON.scalar(args);
  },
  "double": function double(args) {
    return toAminoJSON.scalar(args);
  },
  "float": function float(args) {
    return toAminoJSON.scalar(args);
  },
  bool: function bool(args) {
    return toAminoJSON.scalar(args);
  },
  number: function number(args) {
    return toAminoJSON.scalar(args);
  },
  int32: function int32(args) {
    return toAminoJSON.scalar(args);
  },
  uint32: function uint32(args) {
    return toAminoJSON.scalar(args);
  },
  sint32: function sint32(args) {
    return toAminoJSON.scalar(args);
  },
  fixed32: function fixed32(args) {
    return toAminoJSON.scalar(args);
  },
  sfixed32: function sfixed32(args) {
    return toAminoJSON.scalar(args);
  },
  "long": function long(args) {
    var _getFieldNames2 = (0, _types2.getFieldNames)(args.field),
      propName = _getFieldNames2.propName,
      origName = _getFieldNames2.origName;
    return t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('obj'), t.identifier(origName)), t.conditionalExpression(t.memberExpression(t.identifier('message'), t.identifier(propName)), t.callExpression(t.memberExpression(t.memberExpression(t.identifier('message'), t.identifier(propName)), t.identifier('toString')), []), t.identifier('undefined'))));
  },
  int64: function int64(args) {
    return toAminoJSON["long"](args);
  },
  uint64: function uint64(args) {
    return toAminoJSON["long"](args);
  },
  sint64: function sint64(args) {
    return toAminoJSON["long"](args);
  },
  fixed64: function fixed64(args) {
    return toAminoJSON["long"](args);
  },
  sfixed64: function sfixed64(args) {
    return toAminoJSON["long"](args);
  },
  protoType: function protoType(args) {
    var _getFieldNames3 = (0, _types2.getFieldNames)(args.field),
      propName = _getFieldNames3.propName,
      origName = _getFieldNames3.origName;
    var name = args.context.getTypeName(args.field);
    var defaultValue = t.identifier('undefined');
    if (args.field.type === 'ibc.core.client.v1.Height') {
      defaultValue = t.objectExpression([]);
    }
    return t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('obj'), t.identifier(origName)), t.conditionalExpression(t.memberExpression(t.identifier('message'), t.identifier(propName)), t.callExpression(t.memberExpression(t.identifier(name), t.identifier('toAmino')), [t.memberExpression(t.identifier('message'), t.identifier(propName))]), defaultValue)));
  },
  anyType: function anyType(args) {
    var _getFieldNames4 = (0, _types2.getFieldNames)(args.field),
      propName = _getFieldNames4.propName,
      origName = _getFieldNames4.origName;
    // const typeMap = args.context.store.getTypeUrlMap(args.context.ref);
    // console.log(JSON.stringify(typeMap, null, 2));
    // console.log(JSON.stringify(args.field, null, 2));
    var interfaceName = args.field.options['(cosmos_proto.accepts_interface)'];
    var interfaceFnName = (0, _implements.getInterfaceToAminoName)(interfaceName);
    return t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('obj'), t.identifier(origName)), t.conditionalExpression(t.memberExpression(t.identifier('message'), t.identifier(propName)), t.callExpression(t.identifier(interfaceFnName), [t.tsAsExpression(t.memberExpression(t.identifier('message'), t.identifier(propName)), t.tsTypeReference(t.identifier('Any')))]), t.identifier('undefined'))));
  },
  type: function type(args) {
    if (args.context.options.aminoEncoding.useRecursiveV2encoding == true && args.context.options.interfaces.enabled == true && args.field.type === 'google.protobuf.Any' && args.field.options['(cosmos_proto.accepts_interface)']) {
      return toAminoJSON.anyType(args);
    }
    return toAminoJSON.protoType(args);
  },
  "enum": function _enum(args) {
    return toAminoJSON.scalar(args);
  },
  bytes: function bytes(args) {
    return toAminoJSON.scalar(args);
  },
  duration: function duration(args) {
    return toAminoJSON.type(args);
  },
  timestamp: function timestamp(args) {
    return toAminoJSON.type(args);
  },
  pubkey: function pubkey(args) {
    args.context.addUtil('fromBase64');
    args.context.addUtil('decodeBech32Pubkey');
    var _getFieldNames5 = (0, _types2.getFieldNames)(args.field),
      propName = _getFieldNames5.propName,
      origName = _getFieldNames5.origName;
    return t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('obj'), t.identifier(origName)), t.conditionalExpression(t.memberExpression(t.identifier('message'), t.identifier(propName)),
    //
    t.objectExpression([t.objectProperty(t.identifier('typeUrl'), t.stringLiteral('/cosmos.crypto.secp256k1.PubKey')), t.objectProperty(t.identifier('value'), t.callExpression(t.identifier('fromBase64'), [t.memberExpression(t.callExpression(t.identifier('decodeBech32Pubkey'), [t.memberExpression(t.identifier('message'), t.identifier(propName))]), t.identifier('value'))]))]),
    //
    t.identifier('undefined'))));
  },
  rawBytes: function rawBytes(args) {
    args.context.addUtil('fromUtf8');
    var _getFieldNames6 = (0, _types2.getFieldNames)(args.field),
      propName = _getFieldNames6.propName,
      origName = _getFieldNames6.origName;
    return t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('obj'), t.identifier(origName)), t.conditionalExpression(t.memberExpression(t.identifier('message'), t.identifier(propName)),
    //
    t.callExpression(t.memberExpression(t.identifier('JSON'), t.identifier('parse')), [t.callExpression(t.identifier('fromUtf8'), [t.memberExpression(t.identifier('message'), t.identifier(propName))])]),
    //
    t.identifier('undefined'))));
  },
  wasmByteCode: function wasmByteCode(args) {
    args.context.addUtil('toBase64');
    var _getFieldNames7 = (0, _types2.getFieldNames)(args.field),
      propName = _getFieldNames7.propName,
      origName = _getFieldNames7.origName;
    return t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('obj'), t.identifier(origName)), t.conditionalExpression(t.memberExpression(t.identifier('message'), t.identifier(propName)),
    //
    t.callExpression(t.identifier('toBase64'), [t.memberExpression(t.identifier('message'), t.identifier(propName))]),
    //
    t.identifier('undefined'))));
  },
  keyHash: function keyHash(args) {
    var _getFieldNames8 = (0, _types2.getFieldNames)(args.field),
      propName = _getFieldNames8.propName,
      origName = _getFieldNames8.origName;
    var keyType = args.field.keyType;
    var valueType = args.field.parsedType.name;
    var toAminoJSON = null;
    switch (valueType) {
      case 'string':
        toAminoJSON = t.identifier('v');
        break;
      case 'uint32':
      case 'int32':
        toAminoJSON = t.callExpression(t.memberExpression(t.identifier('Math'), t.identifier('round')), [t.identifier('v')]);
        break;
      case 'int64':
      case 'uint64':
        toAminoJSON = t.callExpression(t.memberExpression(t.identifier('v'), t.identifier('toString')), []);
        break;
      default:
        toAminoJSON = t.callExpression(t.memberExpression(t.identifier(valueType), t.identifier('toAmino')), [t.identifier('v')]);
    }
    return [t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('obj'), t.identifier(origName)), t.objectExpression([]))),
    //
    t.ifStatement(t.memberExpression(t.identifier('message'), t.identifier(propName)), t.blockStatement([t.expressionStatement(t.callExpression(t.memberExpression(t.callExpression(t.memberExpression(t.identifier('Object'), t.identifier('entries')), [t.memberExpression(t.identifier('message'), t.identifier(propName))]), t.identifier('forEach')), [t.arrowFunctionExpression([t.arrayPattern([t.identifier('k'), t.identifier('v')])], t.blockStatement([t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.memberExpression(t.identifier('obj'), t.identifier(origName)), t.identifier('k'), true), toAminoJSON))]))]))]))];
  },
  array: function array(args, expr) {
    var _getFieldNames9 = (0, _types2.getFieldNames)(args.field),
      propName = _getFieldNames9.propName,
      origName = _getFieldNames9.origName;
    return t.ifStatement(t.memberExpression(t.identifier('message'), t.identifier(propName)), t.blockStatement([t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('obj'), t.identifier(origName)), t.callExpression(t.memberExpression(t.memberExpression(t.identifier('message'), t.identifier(propName)), t.identifier('map')), [t.arrowFunctionExpression([t.identifier('e')], expr)])))]), t.blockStatement([t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('obj'), t.identifier(origName)), t.arrayExpression([])))]));
  }
};
exports.toAminoJSON = toAminoJSON;
var arrayTypes = {
  scalar: function scalar() {
    return t.identifier('e');
  },
  string: function string() {
    return arrayTypes.scalar();
  },
  "double": function double() {
    return arrayTypes.scalar();
  },
  "float": function float() {
    return arrayTypes.scalar();
  },
  bool: function bool() {
    return arrayTypes.scalar();
  },
  number: function number() {
    return arrayTypes.scalar();
  },
  int32: function int32() {
    return arrayTypes.number();
  },
  uint32: function uint32() {
    return arrayTypes.number();
  },
  sint32: function sint32() {
    return arrayTypes.number();
  },
  fixed32: function fixed32() {
    return arrayTypes.number();
  },
  sfixed32: function sfixed32() {
    return arrayTypes.number();
  },
  "long": function long(args) {
    return arrayTypes.scalar();
  },
  int64: function int64(args) {
    return arrayTypes["long"](args);
  },
  uint64: function uint64(args) {
    return arrayTypes["long"](args);
  },
  sint64: function sint64(args) {
    return arrayTypes["long"](args);
  },
  fixed64: function fixed64(args) {
    return arrayTypes["long"](args);
  },
  sfixed64: function sfixed64(args) {
    return arrayTypes["long"](args);
  },
  bytes: function bytes(args) {
    return arrayTypes.scalar();
  },
  "enum": function _enum(args) {
    var enumFuncName = args.context.getToEnum(args.field);
    return t.callExpression(t.identifier(enumFuncName), [t.identifier('e')]);
  },
  anyType: function anyType(args) {
    var _getFieldNames10 = (0, _types2.getFieldNames)(args.field),
      propName = _getFieldNames10.propName,
      origName = _getFieldNames10.origName;
    // const typeMap = args.context.store.getTypeUrlMap(args.context.ref);
    // console.log(JSON.stringify(typeMap, null, 2));
    // console.log(JSON.stringify(args.field, null, 2));
    var interfaceName = args.field.options['(cosmos_proto.accepts_interface)'];
    var interfaceFnName = (0, _implements.getInterfaceToAminoName)(interfaceName);
    return t.conditionalExpression(t.identifier('e'), t.callExpression(t.identifier(interfaceFnName), [t.tsAsExpression(t.identifier('e'), t.tsTypeReference(t.identifier('Any')))]), t.identifier('undefined'));
  },
  protoType: function protoType(args) {
    var name = args.context.getTypeName(args.field);
    return t.conditionalExpression(t.identifier('e'), t.callExpression(t.memberExpression(t.identifier(name), t.identifier('toAmino')), [t.identifier('e')]), t.identifier('undefined'));
  },
  type: function type(args) {
    if (args.context.options.aminoEncoding.useRecursiveV2encoding == true && args.context.options.interfaces.enabled == true && args.field.type === 'google.protobuf.Any' && args.field.options['(cosmos_proto.accepts_interface)']) {
      return arrayTypes.anyType(args);
    }
    return arrayTypes.protoType(args);
  }
};
exports.arrayTypes = arrayTypes;
var toAminoMessages = {
  duration: function duration(context, name, proto) {
    return t.returnStatement(t.callExpression(t.memberExpression(t.binaryExpression('+', t.binaryExpression('*', t.callExpression(t.memberExpression(t.memberExpression(t.identifier('message'), t.identifier('seconds')), t.identifier('toInt')), []), _utils.BILLION), t.memberExpression(t.identifier('message'), t.identifier('nanos'))), t.identifier('toString')), []));
  }
};
exports.toAminoMessages = toAminoMessages;
