// osmojs/main/codegen/osmosis/protorev/v1beta1/tx.amino.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AminoConverter = void 0;

var _helpers = require("../../../helpers");

//@ts-nocheck
var AminoConverter = {
  "/osmosis.protorev.v1beta1.MsgSetHotRoutes": {
    aminoType: "osmosis/protorev/set-hot-routes",
    toAmino: function toAmino(_ref) {
      var admin = _ref.admin,
          hotRoutes = _ref.hotRoutes;
      return {
        admin: admin,
        hot_routes: hotRoutes.map(function (el0) {
          return {
            arb_routes: el0.arbRoutes.map(function (el1) {
              return {
                trades: el1.trades.map(function (el2) {
                  return {
                    pool: el2.pool.toString(),
                    token_in: el2.tokenIn,
                    token_out: el2.tokenOut
                  };
                })
              };
            }),
            token_in: el0.tokenIn,
            token_out: el0.tokenOut
          };
        })
      };
    },
    fromAmino: function fromAmino(_ref2) {
      var admin = _ref2.admin,
          hot_routes = _ref2.hot_routes;
      return {
        admin: admin,
        hotRoutes: hot_routes.map(function (el0) {
          return {
            arbRoutes: el0.arb_routes.map(function (el1) {
              return {
                trades: el1.trades.map(function (el2) {
                  return {
                    pool: _helpers.Long.fromString(el2.pool),
                    tokenIn: el2.token_in,
                    tokenOut: el2.token_out
                  };
                })
              };
            }),
            tokenIn: el0.token_in,
            tokenOut: el0.token_out
          };
        })
      };
    }
  },
  "/osmosis.protorev.v1beta1.MsgSetDeveloperAccount": {
    aminoType: "osmosis/protorev/set-developer-account",
    toAmino: function toAmino(_ref3) {
      var admin = _ref3.admin,
          developerAccount = _ref3.developerAccount;
      return {
        admin: admin,
        developer_account: developerAccount
      };
    },
    fromAmino: function fromAmino(_ref4) {
      var admin = _ref4.admin,
          developer_account = _ref4.developer_account;
      return {
        admin: admin,
        developerAccount: developer_account
      };
    }
  }
};
exports.AminoConverter = AminoConverter;
