// osmojs/module/codegen/osmosis/gamm/pool-models/balancer/tx/tx.registry.js
//@ts-nocheck
import { MsgCreateBalancerPool, MsgMigrateSharesToFullRangeConcentratedPosition } from "./tx";
export const registry = [["/osmosis.gamm.poolmodels.balancer.v1beta1.MsgCreateBalancerPool", MsgCreateBalancerPool], ["/osmosis.gamm.poolmodels.balancer.v1beta1.MsgMigrateSharesToFullRangeConcentratedPosition", MsgMigrateSharesToFullRangeConcentratedPosition]];
export const load = protoRegistry => {
  registry.forEach(([typeUrl, mod]) => {
    protoRegistry.register(typeUrl, mod);
  });
};
export const MessageComposer = {
  encoded: {
    createBalancerPool(value) {
      return {
        typeUrl: "/osmosis.gamm.poolmodels.balancer.v1beta1.MsgCreateBalancerPool",
        value: MsgCreateBalancerPool.encode(value).finish()
      };
    },

    migrateSharesToFullRangeConcentratedPosition(value) {
      return {
        typeUrl: "/osmosis.gamm.poolmodels.balancer.v1beta1.MsgMigrateSharesToFullRangeConcentratedPosition",
        value: MsgMigrateSharesToFullRangeConcentratedPosition.encode(value).finish()
      };
    }

  },
  withTypeUrl: {
    createBalancerPool(value) {
      return {
        typeUrl: "/osmosis.gamm.poolmodels.balancer.v1beta1.MsgCreateBalancerPool",
        value
      };
    },

    migrateSharesToFullRangeConcentratedPosition(value) {
      return {
        typeUrl: "/osmosis.gamm.poolmodels.balancer.v1beta1.MsgMigrateSharesToFullRangeConcentratedPosition",
        value
      };
    }

  },
  fromPartial: {
    createBalancerPool(value) {
      return {
        typeUrl: "/osmosis.gamm.poolmodels.balancer.v1beta1.MsgCreateBalancerPool",
        value: MsgCreateBalancerPool.fromPartial(value)
      };
    },

    migrateSharesToFullRangeConcentratedPosition(value) {
      return {
        typeUrl: "/osmosis.gamm.poolmodels.balancer.v1beta1.MsgMigrateSharesToFullRangeConcentratedPosition",
        value: MsgMigrateSharesToFullRangeConcentratedPosition.fromPartial(value)
      };
    }

  }
};
