// osmojs/main/codegen/osmosis/lcd.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createLCDClient = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _lcd = require("@osmonauts/lcd");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var createLCDClient = /*#__PURE__*/function () {
  var _ref2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(_ref) {
    var restEndpoint, requestClient;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            restEndpoint = _ref.restEndpoint;
            requestClient = new _lcd.LCDClient({
              restEndpoint: restEndpoint
            });
            _context.next = 4;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/authz/v1beta1/query.lcd"));
            });

          case 4:
            _context.t0 = _context.sent.LCDQueryClient;
            _context.t1 = {
              requestClient: requestClient
            };
            _context.t2 = new _context.t0(_context.t1);
            _context.t3 = {
              v1beta1: _context.t2
            };
            _context.next = 10;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/bank/v1beta1/query.lcd"));
            });

          case 10:
            _context.t4 = _context.sent.LCDQueryClient;
            _context.t5 = {
              requestClient: requestClient
            };
            _context.t6 = new _context.t4(_context.t5);
            _context.t7 = {
              v1beta1: _context.t6
            };
            _context.next = 16;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/distribution/v1beta1/query.lcd"));
            });

          case 16:
            _context.t8 = _context.sent.LCDQueryClient;
            _context.t9 = {
              requestClient: requestClient
            };
            _context.t10 = new _context.t8(_context.t9);
            _context.t11 = {
              v1beta1: _context.t10
            };
            _context.next = 22;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/gov/v1/query.lcd"));
            });

          case 22:
            _context.t12 = _context.sent.LCDQueryClient;
            _context.t13 = {
              requestClient: requestClient
            };
            _context.t14 = new _context.t12(_context.t13);
            _context.next = 27;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/gov/v1beta1/query.lcd"));
            });

          case 27:
            _context.t15 = _context.sent.LCDQueryClient;
            _context.t16 = {
              requestClient: requestClient
            };
            _context.t17 = new _context.t15(_context.t16);
            _context.t18 = {
              v1: _context.t14,
              v1beta1: _context.t17
            };
            _context.next = 33;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/staking/v1beta1/query.lcd"));
            });

          case 33:
            _context.t19 = _context.sent.LCDQueryClient;
            _context.t20 = {
              requestClient: requestClient
            };
            _context.t21 = new _context.t19(_context.t20);
            _context.t22 = {
              v1beta1: _context.t21
            };
            _context.next = 39;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/tx/v1beta1/service.lcd"));
            });

          case 39:
            _context.t23 = _context.sent.LCDQueryClient;
            _context.t24 = {
              requestClient: requestClient
            };
            _context.t25 = new _context.t23(_context.t24);
            _context.t26 = {
              v1beta1: _context.t25
            };
            _context.next = 45;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/upgrade/v1beta1/query.lcd"));
            });

          case 45:
            _context.t27 = _context.sent.LCDQueryClient;
            _context.t28 = {
              requestClient: requestClient
            };
            _context.t29 = new _context.t27(_context.t28);
            _context.t30 = {
              v1beta1: _context.t29
            };
            _context.t31 = {
              authz: _context.t3,
              bank: _context.t7,
              distribution: _context.t11,
              gov: _context.t18,
              staking: _context.t22,
              tx: _context.t26,
              upgrade: _context.t30
            };
            _context.next = 52;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./concentrated-liquidity/pool-model/query.lcd"));
            });

          case 52:
            _context.t32 = _context.sent.LCDQueryClient;
            _context.t33 = {
              requestClient: requestClient
            };
            _context.t34 = new _context.t32(_context.t33);
            _context.t35 = {
              v1beta1: _context.t34
            };
            _context.next = 58;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./downtime-detector/v1beta1/query.lcd"));
            });

          case 58:
            _context.t36 = _context.sent.LCDQueryClient;
            _context.t37 = {
              requestClient: requestClient
            };
            _context.t38 = new _context.t36(_context.t37);
            _context.t39 = {
              v1beta1: _context.t38
            };
            _context.next = 64;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./epochs/query.lcd"));
            });

          case 64:
            _context.t40 = _context.sent.LCDQueryClient;
            _context.t41 = {
              requestClient: requestClient
            };
            _context.t42 = new _context.t40(_context.t41);
            _context.t43 = {
              v1beta1: _context.t42
            };
            _context.next = 70;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./gamm/v1beta1/query.lcd"));
            });

          case 70:
            _context.t44 = _context.sent.LCDQueryClient;
            _context.t45 = {
              requestClient: requestClient
            };
            _context.t46 = new _context.t44(_context.t45);
            _context.next = 75;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./gamm/v2/query.lcd"));
            });

          case 75:
            _context.t47 = _context.sent.LCDQueryClient;
            _context.t48 = {
              requestClient: requestClient
            };
            _context.t49 = new _context.t47(_context.t48);
            _context.t50 = {
              v1beta1: _context.t46,
              v2: _context.t49
            };
            _context.next = 81;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./ibc-rate-limit/v1beta1/query.lcd"));
            });

          case 81:
            _context.t51 = _context.sent.LCDQueryClient;
            _context.t52 = {
              requestClient: requestClient
            };
            _context.t53 = new _context.t51(_context.t52);
            _context.t54 = {
              v1beta1: _context.t53
            };
            _context.next = 87;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./incentives/query.lcd"));
            });

          case 87:
            _context.t55 = _context.sent.LCDQueryClient;
            _context.t56 = {
              requestClient: requestClient
            };
            _context.t57 = new _context.t55(_context.t56);
            _context.next = 92;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./lockup/query.lcd"));
            });

          case 92:
            _context.t58 = _context.sent.LCDQueryClient;
            _context.t59 = {
              requestClient: requestClient
            };
            _context.t60 = new _context.t58(_context.t59);
            _context.next = 97;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./mint/v1beta1/query.lcd"));
            });

          case 97:
            _context.t61 = _context.sent.LCDQueryClient;
            _context.t62 = {
              requestClient: requestClient
            };
            _context.t63 = new _context.t61(_context.t62);
            _context.t64 = {
              v1beta1: _context.t63
            };
            _context.next = 103;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./pool-incentives/v1beta1/query.lcd"));
            });

          case 103:
            _context.t65 = _context.sent.LCDQueryClient;
            _context.t66 = {
              requestClient: requestClient
            };
            _context.t67 = new _context.t65(_context.t66);
            _context.t68 = {
              v1beta1: _context.t67
            };
            _context.next = 109;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./poolmanager/v1beta1/query.lcd"));
            });

          case 109:
            _context.t69 = _context.sent.LCDQueryClient;
            _context.t70 = {
              requestClient: requestClient
            };
            _context.t71 = new _context.t69(_context.t70);
            _context.t72 = {
              v1beta1: _context.t71
            };
            _context.next = 115;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./protorev/v1beta1/query.lcd"));
            });

          case 115:
            _context.t73 = _context.sent.LCDQueryClient;
            _context.t74 = {
              requestClient: requestClient
            };
            _context.t75 = new _context.t73(_context.t74);
            _context.t76 = {
              v1beta1: _context.t75
            };
            _context.next = 121;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./superfluid/query.lcd"));
            });

          case 121:
            _context.t77 = _context.sent.LCDQueryClient;
            _context.t78 = {
              requestClient: requestClient
            };
            _context.t79 = new _context.t77(_context.t78);
            _context.next = 126;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./tokenfactory/v1beta1/query.lcd"));
            });

          case 126:
            _context.t80 = _context.sent.LCDQueryClient;
            _context.t81 = {
              requestClient: requestClient
            };
            _context.t82 = new _context.t80(_context.t81);
            _context.t83 = {
              v1beta1: _context.t82
            };
            _context.next = 132;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./twap/v1beta1/query.lcd"));
            });

          case 132:
            _context.t84 = _context.sent.LCDQueryClient;
            _context.t85 = {
              requestClient: requestClient
            };
            _context.t86 = new _context.t84(_context.t85);
            _context.t87 = {
              v1beta1: _context.t86
            };
            _context.next = 138;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./txfees/v1beta1/query.lcd"));
            });

          case 138:
            _context.t88 = _context.sent.LCDQueryClient;
            _context.t89 = {
              requestClient: requestClient
            };
            _context.t90 = new _context.t88(_context.t89);
            _context.t91 = {
              v1beta1: _context.t90
            };
            _context.next = 144;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./valset-pref/v1beta1/query.lcd"));
            });

          case 144:
            _context.t92 = _context.sent.LCDQueryClient;
            _context.t93 = {
              requestClient: requestClient
            };
            _context.t94 = new _context.t92(_context.t93);
            _context.t95 = {
              v1beta1: _context.t94
            };
            _context.t96 = {
              concentratedliquidity: _context.t35,
              downtimedetector: _context.t39,
              epochs: _context.t43,
              gamm: _context.t50,
              ibcratelimit: _context.t54,
              incentives: _context.t57,
              lockup: _context.t60,
              mint: _context.t64,
              poolincentives: _context.t68,
              poolmanager: _context.t72,
              protorev: _context.t76,
              superfluid: _context.t79,
              tokenfactory: _context.t83,
              twap: _context.t87,
              txfees: _context.t91,
              valsetpref: _context.t95
            };
            return _context.abrupt("return", {
              cosmos: _context.t31,
              osmosis: _context.t96
            });

          case 150:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function createLCDClient(_x) {
    return _ref2.apply(this, arguments);
  };
}();

exports.createLCDClient = createLCDClient;
