// osmojs/main/codegen/osmosis/rpc.tx.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createRPCMsgClient = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var createRPCMsgClient = /*#__PURE__*/function () {
  var _ref2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(_ref) {
    var rpc;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            rpc = _ref.rpc;
            _context.next = 3;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/authz/v1beta1/tx.rpc.msg"));
            });

          case 3:
            _context.t0 = _context.sent.MsgClientImpl;
            _context.t1 = rpc;
            _context.t2 = new _context.t0(_context.t1);
            _context.t3 = {
              v1beta1: _context.t2
            };
            _context.next = 9;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/bank/v1beta1/tx.rpc.msg"));
            });

          case 9:
            _context.t4 = _context.sent.MsgClientImpl;
            _context.t5 = rpc;
            _context.t6 = new _context.t4(_context.t5);
            _context.t7 = {
              v1beta1: _context.t6
            };
            _context.next = 15;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/distribution/v1beta1/tx.rpc.msg"));
            });

          case 15:
            _context.t8 = _context.sent.MsgClientImpl;
            _context.t9 = rpc;
            _context.t10 = new _context.t8(_context.t9);
            _context.t11 = {
              v1beta1: _context.t10
            };
            _context.next = 21;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/gov/v1/tx.rpc.msg"));
            });

          case 21:
            _context.t12 = _context.sent.MsgClientImpl;
            _context.t13 = rpc;
            _context.t14 = new _context.t12(_context.t13);
            _context.next = 26;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/gov/v1beta1/tx.rpc.msg"));
            });

          case 26:
            _context.t15 = _context.sent.MsgClientImpl;
            _context.t16 = rpc;
            _context.t17 = new _context.t15(_context.t16);
            _context.t18 = {
              v1: _context.t14,
              v1beta1: _context.t17
            };
            _context.next = 32;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/staking/v1beta1/tx.rpc.msg"));
            });

          case 32:
            _context.t19 = _context.sent.MsgClientImpl;
            _context.t20 = rpc;
            _context.t21 = new _context.t19(_context.t20);
            _context.t22 = {
              v1beta1: _context.t21
            };
            _context.next = 38;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("../cosmos/upgrade/v1beta1/tx.rpc.msg"));
            });

          case 38:
            _context.t23 = _context.sent.MsgClientImpl;
            _context.t24 = rpc;
            _context.t25 = new _context.t23(_context.t24);
            _context.t26 = {
              v1beta1: _context.t25
            };
            _context.t27 = {
              authz: _context.t3,
              bank: _context.t7,
              distribution: _context.t11,
              gov: _context.t18,
              staking: _context.t22,
              upgrade: _context.t26
            };
            _context.next = 45;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./concentrated-liquidity/tx.rpc.msg"));
            });

          case 45:
            _context.t28 = _context.sent.MsgClientImpl;
            _context.t29 = rpc;
            _context.t30 = new _context.t28(_context.t29);
            _context.t31 = {
              v1beta1: _context.t30
            };
            _context.next = 51;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./gamm/pool-models/balancer/tx/tx.rpc.msg"));
            });

          case 51:
            _context.t32 = _context.sent.MsgClientImpl;
            _context.t33 = rpc;
            _context.t34 = new _context.t32(_context.t33);
            _context.t35 = {
              v1beta1: _context.t34
            };
            _context.next = 57;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./gamm/pool-models/stableswap/tx.rpc.msg"));
            });

          case 57:
            _context.t36 = _context.sent.MsgClientImpl;
            _context.t37 = rpc;
            _context.t38 = new _context.t36(_context.t37);
            _context.t39 = {
              v1beta1: _context.t38
            };
            _context.t40 = {
              balancer: _context.t35,
              stableswap: _context.t39
            };
            _context.next = 64;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./gamm/v1beta1/tx.rpc.msg"));
            });

          case 64:
            _context.t41 = _context.sent.MsgClientImpl;
            _context.t42 = rpc;
            _context.t43 = new _context.t41(_context.t42);
            _context.t44 = {
              poolmodels: _context.t40,
              v1beta1: _context.t43
            };
            _context.next = 70;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./incentives/tx.rpc.msg"));
            });

          case 70:
            _context.t45 = _context.sent.MsgClientImpl;
            _context.t46 = rpc;
            _context.t47 = new _context.t45(_context.t46);
            _context.next = 75;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./lockup/tx.rpc.msg"));
            });

          case 75:
            _context.t48 = _context.sent.MsgClientImpl;
            _context.t49 = rpc;
            _context.t50 = new _context.t48(_context.t49);
            _context.next = 80;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./poolmanager/v1beta1/tx.rpc.msg"));
            });

          case 80:
            _context.t51 = _context.sent.MsgClientImpl;
            _context.t52 = rpc;
            _context.t53 = new _context.t51(_context.t52);
            _context.t54 = {
              v1beta1: _context.t53
            };
            _context.next = 86;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./protorev/v1beta1/tx.rpc.msg"));
            });

          case 86:
            _context.t55 = _context.sent.MsgClientImpl;
            _context.t56 = rpc;
            _context.t57 = new _context.t55(_context.t56);
            _context.t58 = {
              v1beta1: _context.t57
            };
            _context.next = 92;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./superfluid/tx.rpc.msg"));
            });

          case 92:
            _context.t59 = _context.sent.MsgClientImpl;
            _context.t60 = rpc;
            _context.t61 = new _context.t59(_context.t60);
            _context.next = 97;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./tokenfactory/v1beta1/tx.rpc.msg"));
            });

          case 97:
            _context.t62 = _context.sent.MsgClientImpl;
            _context.t63 = rpc;
            _context.t64 = new _context.t62(_context.t63);
            _context.t65 = {
              v1beta1: _context.t64
            };
            _context.next = 103;
            return Promise.resolve().then(function () {
              return _interopRequireWildcard(require("./valset-pref/v1beta1/tx.rpc.msg"));
            });

          case 103:
            _context.t66 = _context.sent.MsgClientImpl;
            _context.t67 = rpc;
            _context.t68 = new _context.t66(_context.t67);
            _context.t69 = {
              v1beta1: _context.t68
            };
            _context.t70 = {
              concentratedliquidity: _context.t31,
              gamm: _context.t44,
              incentives: _context.t47,
              lockup: _context.t50,
              poolmanager: _context.t54,
              protorev: _context.t58,
              superfluid: _context.t61,
              tokenfactory: _context.t65,
              valsetpref: _context.t69
            };
            return _context.abrupt("return", {
              cosmos: _context.t27,
              osmosis: _context.t70
            });

          case 109:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function createRPCMsgClient(_x) {
    return _ref2.apply(this, arguments);
  };
}();

exports.createRPCMsgClient = createRPCMsgClient;
